### DISCUSSION TITLE
[Android] Java.Lang.RuntimeException: Canvas: trying to draw too large bitmap.

### DISCUSSION THREAD

**Discussion by greg84** [xamarin/Xamarin.Forms#14978] [Q&A] [answered]
I'm seeing this exception logged in production since updating an app with a redesign and an update from Xamarin Forms 5.0.0.2012 to 5.0.0.2244.

`Java.Lang.RuntimeException: Canvas: trying to draw too large(198450000bytes) bitmap.`

It's affecting so far about 1% of Android users, causing a crash, and I'm unable to reproduce it because it seems to crash after a few seconds and only some of the time.

The first page that's displayed is a CarouselView with a few images for onboarding, but none of them exceed a couple of hundred KB, so it doesn't explain the error.

Any ideas what I could try to work out what's going on?

In the stack trace (below) it mentions ShellFlyoutRenderer, but I'm not using a flyout, only a tab bar, which is weird.

<details>
  <summary>Full stack trace from App Center</summary>

  ```
JniEnvironment+InstanceMethods.CallNonvirtualVoidMethod (Java.Interop.JniObjectReference instance, Java.Interop.JniObjectReference type, Java.Interop.JniMethodInfo method, Java.Interop.JniArgumentValue* args)
JniPeerMembers+JniInstanceMethods.InvokeVirtualVoidMethod (System.String encodedMember, Java.Interop.IJavaPeerable self, Java.Interop.JniArgumentValue* parameters)
View.Draw (Android.Graphics.Canvas canvas)
ImageRenderer.Draw (Android.Graphics.Canvas canvas)
View.n_Draw_Landroid_graphics_Canvas_ (System.IntPtr jnienv, System.IntPtr native__this, System.IntPtr native_canvas)
(wrapper dynamic-method) Android.Runtime.DynamicMethodNameCounter.83(intptr,intptr,intptr)
java.lang.RuntimeException: Canvas: trying to draw too large(115200000bytes) bitmap.
android.view.DisplayListCanvas.throwIfCannotDraw DisplayListCanvas.java:229
android.view.RecordingCanvas.drawBitmap RecordingCanvas.java:97
android.graphics.drawable.BitmapDrawable.draw BitmapDrawable.java:529
android.widget.ImageView.onDraw ImageView.java:1370
android.view.View.draw View.java:19382
crc64ee486da937c010f4.ImageRenderer.n_draw(Native Method)
crc64ee486da937c010f4.ImageRenderer.draw ImageRenderer.java:64
android.view.View.updateDisplayListIfDirty View.java:18311
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.updateDisplayListIfDirty View.java:18297
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.updateDisplayListIfDirty View.java:18297
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.updateDisplayListIfDirty View.java:18297
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.draw View.java:19385
crc643f46942d9dd1fff9.VisualElementRenderer_1.n_draw(Native Method)
crc643f46942d9dd1fff9.VisualElementRenderer_1.draw VisualElementRenderer_1.java:100
android.view.View.updateDisplayListIfDirty View.java:18311
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.updateDisplayListIfDirty View.java:18297
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.draw View.java:19385
androidx.viewpager.widget.ViewPager.draw ViewPager.java:2426
android.view.View.updateDisplayListIfDirty View.java:18311
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
androidx.coordinatorlayout.widget.CoordinatorLayout.drawChild CoordinatorLayout.java:1277
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.updateDisplayListIfDirty View.java:18297
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.updateDisplayListIfDirty View.java:18297
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.updateDisplayListIfDirty View.java:18297
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.updateDisplayListIfDirty View.java:18297
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
androidx.drawerlayout.widget.DrawerLayout.drawChild DrawerLayout.java:1478
crc643f46942d9dd1fff9.ShellFlyoutRenderer.n_drawChild(Native Method)
crc643f46942d9dd1fff9.ShellFlyoutRenderer.drawChild ShellFlyoutRenderer.java:59
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.updateDisplayListIfDirty View.java:18297
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.updateDisplayListIfDirty View.java:18297
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.updateDisplayListIfDirty View.java:18297
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.updateDisplayListIfDirty View.java:18297
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.updateDisplayListIfDirty View.java:18297
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.updateDisplayListIfDirty View.java:18297
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.updateDisplayListIfDirty View.java:18297
android.view.View.draw View.java:19099
android.view.ViewGroup.drawChild ViewGroup.java:4401
android.view.ViewGroup.dispatchDraw ViewGroup.java:4180
android.view.View.draw View.java:19385
com.android.internal.policy.DecorView.draw DecorView.java:1007
android.view.View.updateDisplayListIfDirty View.java:18311
android.view.ThreadedRenderer.updateViewTreeDisplayList ThreadedRenderer.java:676
android.view.ThreadedRenderer.updateRootDisplayList ThreadedRenderer.java:682
android.view.ThreadedRenderer.draw ThreadedRenderer.java:790
android.view.ViewRootImpl.draw ViewRootImpl.java:3091
android.view.ViewRootImpl.performDraw ViewRootImpl.java:2884
android.view.ViewRootImpl.performTraversals ViewRootImpl.java:2437
android.view.ViewRootImpl.doTraversal ViewRootImpl.java:1446
android.view.ViewRootImpl$TraversalRunnable.run ViewRootImpl.java:6954
android.view.Choreographer$CallbackRecord.run Choreographer.java:980
android.view.Choreographer.doCallbacks Choreographer.java:792
android.view.Choreographer.doFrame Choreographer.java:724
android.view.Choreographer$FrameDisplayEventReceiver.run Choreographer.java:966
android.os.Handler.handleCallback Handler.java:790
android.os.Handler.dispatchMessage Handler.java:99
android.os.Looper.loop Looper.java:192
android.app.ActivityThread.main ActivityThread.java:6711
java.lang.reflect.Method.invoke(Native Method)
com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run RuntimeInit.java:549
com.android.internal.os.ZygoteInit.main ZygoteInit.java:826
```

</details>

**Comment by jfversluis** [comment]
I'm not convinced think this is a bug on our side, it mentions it tries to load a 115 megabyte image, so somehow users can end up doing that. Do you let people pick images or something like that?

It is weird that it mentions the flyout... You do use Shell, just not the flyout?

**Reply by greg84** [reply]
Thanks @jfversluis, some more details:

Users are not able to upload their own images.

All of the images included in the application package are no bigger than a few hundred kb.

The app mostly crashes within the first few seconds, suggesting an issue during startup, sometimes within 1 or 2 seconds, so it could be crashing before the user has interacted with it. 

You are correct - I am using Shell but not the flyout. I actually used to use a Flyout, and that version did not crash, I since changed to a tab bar (still using shell) with no flyout and have now started to see these errors reported.

I've purchased one of the devices that's reporting the crash from eBay (it's a 5 year old phone) to try and reproduce it!

This is the line it's crashing on: https://github.com/xamarin/Xamarin.Forms/blob/caab66bcf9614aca0c0805d560a34e176d196e17/Xamarin.Forms.Platform.Android/Renderers/ShellFlyoutRenderer.cs#L151

Do you know what image it's trying to draw at this point? Especially when no flyout is being defined and I'm explicitly setting `Shell.FlyoutBehavior` to `Disabled`.

**Reply by greg84** [reply]
I was able to reproduce the issue on the device I purchased.

I had to use a drawable-xxhdpi folder for an image that was being displayed. Originally the image was in the drawable folder and Android was trying to scale it up by a ridiculous amount.

Thanks for your help on this @jfversluis.

**Comment by greg84** [comment, marked as answer]
I had to use a drawable-xxhdpi folder for an image that was being displayed. Originally the image was in the drawable folder and Android was trying to scale it up by a ridiculous amount.

**Reply by sirqoharastus** [reply]
How were you able to identify the particular image causing the issue?

**Comment by smalgin** [comment]
Hello all. 
I am victim of the same problem, but worse, as I don't have a static file.

I download a thumbnail dynamically into a temporary file and specify ImageSource like this: 
-
 ThumbSources.Add(Xamarin.Forms.ImageSource.FromFile(thumbPath));

How can I specify 'xxdpi' option in my code to make Xamarin ImageView work again?
My only tester has a device where this bug appears. Swiping it under the rug is NOT an option.

**Reply by aidan-bailey** [reply]
I am having this issue in Maui, but worse, as **all** the images within my application are fetched from a server during startup haha. Not rug sweepable indeed.

**Reply by aidan-bailey** [reply]
I found a solution perfect for my specific situation: disabling Android scaling entirely by setting `anyDensity` to `false` in Platforms/Android/AndroidManifest.xml. We manage all image scaling anyway. I imagine this is not a solution applicable to the general case.

**Comment by kiddailey** [comment]
Just to add an additional detail to anyone stumbling here.  This also occurs with static images in "drawable" and older API levels and can be triggered in emulators consistently.  For example, a small 13kb PNG on API 33 device will work fine, but on API 26 for some reason the image is scaled insanely large.

Moving the images into the mipmap folders is one solution, but requires a bit more work.

Can also confirm that adding the android:anyDensity property to the manifest does indeed work around the issue, though it could have side effects.

```
<manifest ... >
    <supports-screens 
        android:anyDensity="false" 
        android:xlargeScreens="true" 
        android:largeScreens="true" 
        android:normalScreens="true" 
        android:smallScreens="true" />
</manifest>
```
